{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################\n",
    "# Filtering and normalization of ASV count \n",
    "##########################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. TRUE\n",
       "2. TRUE\n",
       "3. TRUE\n",
       "4. TRUE\n",
       "5. TRUE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] TRUE\n",
       "\n",
       "[[2]]\n",
       "[1] TRUE\n",
       "\n",
       "[[3]]\n",
       "[1] TRUE\n",
       "\n",
       "[[4]]\n",
       "[1] TRUE\n",
       "\n",
       "[[5]]\n",
       "[1] TRUE\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Set seed\n",
    "set.seed(32426)\n",
    "\n",
    "# Load libraries\n",
    "source(\"scr/functions/general/load_abs_install_pkg.R\")\n",
    "\n",
    "load_abs_install_pkg(c(\"phyloseq\", \"metagenomeSeq\", \"tidyverse\", \"RColorBrewer\", \"gmodels\"))\n",
    "\n",
    "# Create output directory \n",
    "dir.create(\"output/3_filt_norm_phyl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 1. Read in phyloseq object generated by DADA2\n",
    "###############################################\n",
    "\n",
    "ps.0 <- readRDS(\"output/2_dada2/phyloseq0.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 2. Remove ASVs that have count or/end prevalence less than 3 \n",
    "###########################################################\n",
    "\n",
    "# Remove ASVs with less than 3 observations   \n",
    "ps.0f <- prune_taxa(taxa = colSums(ps.0@otu_table) > 3 , x= ps.0) \n",
    "\n",
    "# Extract otu table into otu_prev object\n",
    "otu_prev <- otu_table(ps.0f)\n",
    "\n",
    "# Convert otu_prev from count to prevalence data \n",
    "otu_prev[otu_prev > 1] <- 1\n",
    "\n",
    "# Remove taxa with prevalence less than 3 \n",
    "ps_tf1 <- prune_taxa(taxa = colSums(otu_prev) > 3 , x= ps.0f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 3. Remove sample that have less than a 1000 observations  \n",
    "######################################################\n",
    "\n",
    "# Extract ASV table into otu_prev object\n",
    "o.tab.ps_tf1 <- otu_table(ps_tf1)\n",
    "\n",
    "# Prune samples with less than 1000 observation from the phyloseq object \n",
    "ps_tf2 <- prune_samples(samples = rownames(o.tab.ps_tf1)[rowSums(o.tab.ps_tf1) > 1000], ps_tf1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Default value being used.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# 4. Normalize ASVs abundance using CSS as implemented in the metagenomSeq \n",
    "##########################################################################\n",
    "\n",
    "# Prepare data from phyloseq package \n",
    "mg.ps_tf2 <- phyloseq_to_metagenomeSeq(ps_tf2)\n",
    "\n",
    "# Calculate cumulative statistics \n",
    "p <- metagenomeSeq::cumNormStatFast(mg.ps_tf2)\n",
    "\n",
    "# Normalize count \n",
    "mg.c.ps_tf2 <- metagenomeSeq::cumNorm(mg.ps_tf2, p = p)\n",
    "\n",
    "# Convert metagenomSeq object with normalized count into a dataframe  \n",
    "css.otu.all <- data.frame(otu_table(metagenomeSeq::MRcounts(mg.c.ps_tf2, \n",
    "                        norm = TRUE, log = TRUE), taxa_are_rows = FALSE))\n",
    "\n",
    "# Convert into matrix \n",
    "css.otu.all.m <- as.matrix(t(css.otu.all))\n",
    "\n",
    "# Replace \".\" with \"-\" in rownames \n",
    "rname <- gsub(\"[.]\", \"-\", rownames(css.otu.all.m))\n",
    "\n",
    "# Replace remove \"X\" from rownames \n",
    "rownames(css.otu.all.m) <- gsub(\"X\", \"\", rname)\n",
    "\n",
    "# Make a copy of the original phyloseq object \n",
    "ps_tf2_css <- ps_tf2\n",
    "\n",
    "# Replace ASV count table with CSS normalized ASV table \n",
    "ps_tf2_css@otu_table@.Data <- css.otu.all.m \n",
    "\n",
    "# Write new phyloseq into RDS file \n",
    "saveRDS(ps_tf2_css, \"output/3_filt_norm_phyl/ps_tf2_css.RDS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 5. Extract general information about sequencing\n",
    "##################################################\n",
    "\n",
    "# Median number of samples per animal \n",
    "med.samp.n <- median(table(ps_tf2_css@sam_data$CowN))\n",
    "\n",
    "# Write information about number of sample per animal into a file \n",
    "write_csv(x = as.data.frame(table(ps_tf2_css@sam_data$CowN)),\n",
    "          path = \"output/3_filt_norm_phyl/Table_S1.csv\")\n",
    "\n",
    "# Extract otu table from filtered and not normalized phyloseq object for \n",
    "#         summary statistics \n",
    "otus.tab <- otu_table(ps_tf2)\n",
    "\n",
    "# Total observations  \n",
    "reads.sum <- sum(otus.tab)\n",
    "\n",
    "# Observations Median\n",
    "reads.med <- median(rowSums(otus.tab))\n",
    "\n",
    "# Observations Min \n",
    "reads.min <- min(rowSums(otus.tab))\n",
    "\n",
    "# Observations Max \n",
    "reads.max <- max(rowSums(otus.tab))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 6. Visualize observations distribution \n",
    "########################################\n",
    "\n",
    "# Calculate number of observations per sample \n",
    "reads.pl.df <- as.data.frame(rowSums(otus.tab))\n",
    "\n",
    "# Adjust column names \n",
    "colnames(reads.pl.df) <- \"Reads\"\n",
    "\n",
    "# Add samples ID column \n",
    "reads.pl.df$SampleID <- rownames(reads.pl.df)\n",
    "\n",
    "# Order samples by number of observations \n",
    "reads.pl.df <- reads.pl.df[order(reads.pl.df$Reads), ]\n",
    "\n",
    "# Order levels of samples\n",
    "reads.pl.df$SampleID <- factor(reads.pl.df$SampleID, levels=unique(reads.pl.df$SampleID))\n",
    "\n",
    "# Plot number of observations per sample\n",
    "reads.pers.p <- ggplot(reads.pl.df, aes(x = SampleID, y = Reads)) + \n",
    "  geom_bar(stat=\"identity\") + \n",
    "  theme_bw() + \n",
    "  geom_hline(yintercept = mean(reads.pl.df$Reads), color=\"blue\") +\n",
    "  theme(axis.text.x = element_blank()) + \n",
    "  ylab(\"Reads number\") + \n",
    "  xlab(\"Samples\") + \n",
    "  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())\n",
    "\n",
    "# Save the plot\n",
    "ggsave(filename = \"output/3_filt_norm_phyl/Figure_S2.pdf\", \n",
    "       plot = reads.pers.p, width = 10, height = 6)\n",
    "ggsave(filename = \"output/3_filt_norm_phyl//Figure_S2.jpg\", \n",
    "       plot = reads.pers.p, width = 10, height = 6, dpi = 400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 7. Visualization of phylogenetic composition on Phylum level \n",
    "##############################################################\n",
    "\n",
    "# Prepare data for plotting \n",
    "# Glom phyloseq to Phylum level \n",
    "ps.tf2.plot <- tax_glom(ps_tf2, taxrank = \"Phylum\")\n",
    "\n",
    "# Transform count to relative abundance \n",
    "ps.tf2.plot <- transform_sample_counts(ps.tf2.plot, function(x) x / sum(x) * 100)\n",
    "\n",
    "# Extract otu table \n",
    "phy.plot.d <- data.frame(t(otu_table(ps.tf2.plot)))\n",
    "\n",
    "# Calculate row sums \n",
    "phy.plot.rs <- rowSums(phy.plot.d)\n",
    "\n",
    "# Add column with Phylum names \n",
    "phy.plot.d$Phylum <- as.character(tax_table(ps.tf2.plot)[,\"Phylum\"])\n",
    "                                       \n",
    "# Melt dataframe into long format \n",
    "phy.plot.dm <- gather(phy.plot.d, SampleID, Abundance, -Phylum)\n",
    "                                       \n",
    "# Order factors in column Phylum by abundance \n",
    "phy.plot.dm$Phylum <- factor(phy.plot.dm$Phylum, \n",
    "                             levels = c(phy.plot.dm$Phylum[order(phy.plot.rs, decreasing = TRUE)]))\n",
    "\n",
    "# Prepare custom colors pallet \n",
    "plot.col <- c(brewer.pal(n = 8, name = \"Dark2\"), replicate('#666666', n = 7))\n",
    "\n",
    "# Plot phylogenetic composition on Phylum level\n",
    "phy.plot <- ggplot(phy.plot.dm, aes(x = SampleID, y = Abundance, fill = Phylum)) + \n",
    "        geom_bar(stat = \"identity\") + \n",
    "        scale_fill_manual(values = plot.col) + \n",
    "        theme_bw() + \n",
    "        theme(axis.text.x = element_blank(), \n",
    "              panel.grid = element_blank(), \n",
    "              axis.ticks.x = element_blank(), )\n",
    "\n",
    "# Save the plot\n",
    "ggsave(filename = \"output/3_filt_norm_phyl/Figure_S3.pdf\", width = 10, height = 5.5)\n",
    "ggsave(filename = \"output/3_filt_norm_phyl/Figure_S3.jpg\", width = 10, height = 5.5, dpi = 400)\n",
    "                                       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”Warning message in ci.numeric(newX[, i], ...):\n",
      "“No class or unkown class.  Using default calcuation.”"
     ]
    }
   ],
   "source": [
    "\n",
    "# 8. Summary table of phylogenetic composition on Phylum level \n",
    "##############################################################\n",
    "\n",
    "# Add phyla names as row names \n",
    "rownames(phy.plot.d) <- phy.plot.d$Phylum\n",
    "\n",
    "# Remove the Phylum column \n",
    "phy.plot.d2 <- phy.plot.d[, ! colnames(phy.plot.d) %in% \"Phylum\"]\n",
    "\n",
    "# Make summary statistics of the phyla relative abundance\n",
    "sum.t <- as.data.frame(summary(t(phy.plot.d2)))\n",
    "\n",
    "# Format summary statistics of the phyla relative abundance\n",
    "sum.t2<- as.data.frame(cbind(as.character(sum.t$Var2), \n",
    "              str_split(sum.t$Freq, pattern = \":\", simplify = TRUE)))\n",
    "\n",
    "# Convert Summary statistics into a wide table\n",
    "sum.t3 <- spread(sum.t2, V2, V3)\n",
    "\n",
    "# Reorder Summary statistics table for future bind with CI calculation results \n",
    "sum.t3 <- sum.t3[order(trimWhiteSpace(as.character(sum.t3$V1))), ]\n",
    "\n",
    "# Calculate CI per phylum \n",
    "ci.t3 <- round(t(apply(phy.plot.d2, 1, ci)), 2)\n",
    "\n",
    "# Reorder CI table to further combine with Summary statistics \n",
    "ci.t3 <- ci.t3[order(rownames(ci.t3)), ]\n",
    "\n",
    "# Bind (wide) CI and Summary tables \n",
    "summ.save <- cbind(sum.t3,  ci.t3)\n",
    "\n",
    "# Write the combined table into file \n",
    "write_csv(x = summ.save, path = \"output/3_filt_norm_phyl/Table_S2.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
